### PR TITLE
Rewrite algo::diff to support insertion and deletion

### PR DESCRIPTION
**Pull request by Veykril**
This in turn also makes `algo::diff` generate finer diffs(maybe even minimal diffs?) as insertions and deletions aren't always represented as as replacements of parent nodes now.

Required for #6287 to go on.

### CODE REVIEW

File: crates/ide/src/diagnostics.rs [hunk 1]
```diff
@@ -613,7 +613,7 @@ fn main() {
 pub struct Foo { pub a: i32, pub b: i32 }
 "#,
             r#"
-fn {a:42, b: ()} {}
+fn some(, b: ()} {}
 fn items() {}
```
Comment (Veykril) [inline crates/ide/src/diagnostics.rs:617]: Thanks to this macro text expecting already broken output you can see the result of these changes in that the diff became smaller here 😄

File: crates/syntax/src/algo.rs [hunk 2]
```diff
@@ -108,40 +110,53 @@ pub enum InsertPosition<T> {
 
 pub struct TreeDiff {
     replacements: FxHashMap<SyntaxElement, SyntaxElement>,
+    deletions: Vec<SyntaxElement>,
+    // the vec as well as the indexmap are both here to preserve order
+    insertions:
+        IndexMap<SyntaxElement, Vec<SyntaxElement>, BuildHasherDefault<rustc_hash::FxHasher>>,
```
Comment (Veykril) [inline crates/syntax/src/algo.rs:116]: Insertion order is important as insertions might be chained after each other. The `IndexMap` could be replace by `Vec<(SyntaxElement, Vec<SyntaxElement>)` with binary_search insertion on the first tuple part as the key but I'm uncertain what would be better.

Comment (matklad) [inline crates/syntax/src/algo.rs:116]: Minor, but let's `type FxIndexMap = `

File: crates/syntax/src/algo.rs [hunk 3]
```diff
@@ -150,18 +165,43 @@ pub fn diff(from: &SyntaxNode, to: &SyntaxNode) -> TreeDiff {
...
+                        break;
+                    }
+                },
+                (Some(element), None) => {
+                    diff.deletions.push(element);
+                }
+                (Some(ref lhs_ele), Some(ref rhs_ele)) if syntax_element_eq(lhs_ele, rhs_ele) => {}
+                (Some(lhs_ele), Some(rhs_ele)) => go(diff, lhs_ele, rhs_ele),
+            }
+            last_lhs = lhs_child.or(last_lhs);
```
Comment (matklad) [inline crates/syntax/src/algo.rs:203]: Hm, would this work corretly if the `go` ends up replacing the whole `lhs_child`, and then, on the next iteration, we use it as an anchor in inserts?

Comment (matklad) [reply crates/syntax/src/algo.rs:203]: Ah, it seems it will work, cool!

File: crates/syntax/src/algo.rs [hunk 4]
```diff
@@ -150,18 +165,43 @@ pub fn diff(from: &SyntaxNode, to: &SyntaxNode) -> TreeDiff {
...
+        loop {
+            let lhs_child = lhs_children.next();
+            match (lhs_child.clone(), rhs_children.next()) {
+                (None, None) => break,
+                (None, Some(element)) => match last_lhs.clone() {
+                    Some(prev) => {
+                        diff.insertions.entry(prev).or_insert_with(Vec::new).push(element);
+                    }
+                    // first iteration, this means we got no anchor element to insert after
+                    // therefor replace the parent node instead
```
Comment (matklad) [inline crates/syntax/src/algo.rs:191]: An alternative is to key insertion not by `SyntaxElement`, but by `(SyntaxElement, bool)`, where bool should choose between prev sibling and parent. But this is also ok

**Comment by matklad** [comment]
Nifty, didn't realize that we can improve tree diffing this way!

r=me wiith FxIndexMap nit fixed

If you feel like ii though, adding some unit-tests for tree diff based on `expect_test` might be useufl long-term

**Pull request merged**